---
title: "0b_species_list_compare"
output: html_document
---

```{r}
# setup

# minimal libraries
librarian::shelf(tidyverse, dbplyr, RPostgres, DBI, RIBBiTR-BII/ribbitrrr, yaml, here, phylotools, Biostrings, pwalign, msa)

# read in config files
config_12s = read_yaml(here("16S Process TEST", "runs/methods_2026-01-12/12S/output/metadata/config.yml"))
config_16s = read_yaml(here("16S Process TEST", "runs/methods_2026-01-12/16S/output/metadata/config.yml"))

# establish database connection
dbcon = hopToDB("ribbitr")

# download RIBBiTR species list, filter to amphibians
taxa_rib = tbl(dbcon, Id("survey_data", "taxonomy")) %>%
  filter(itis_rank == "species",
         itis_class == "Amphibia") %>%
  select(taxon_id,
         itis_tsn,
         itis_order,
         itis_family,
         itis_genus,
         itis_species,
         ncbi_id,
         ncbi_taxon) %>%
  collect() %>%
  mutate(sn_rib = coalesce(ncbi_taxon, itis_species)) %>%
  separate(sn_rib, c("ncbi_genus", "ncbi_species"), sep = " ", remove = FALSE)

# check for discrepancies between itis and ncbi
itis_ncbi_dif = taxa_rib %>%
  filter(itis_species != ncbi_taxon)

# read in 12s taxonomies, filter to amphibians
taxa_12s_raw = read.table(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_taxa_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_taxa_extracted"))), 
                    "/data/taxonomy.tsv")),
             header = TRUE, 
             sep = '\t',
             quote = '') %>%
  separate(Taxon, into = c("k", "p", "c", "o", "f", "g", "s"), sep = ";") %>%
  mutate(across(everything(), ~ str_remove(., "^[a-z]__"))) %>%
  filter(c == "Amphibia",
         s != "")

# rename columns
colnames(taxa_12s_raw) = c("feature_id",
                          "kingdom",
                          "phylum",
                          "class",
                          "order",
                          "family",
                          "genus",
                          "species")

# count by taxonomy
taxa_12s = taxa_12s_raw %>%
  group_by(order, family, genus, species) %>%
  summarise(n_12s = n(),
            .groups = "drop") %>%
  mutate(sn_12s = paste(genus, species))

# read in 16s taxonomy, filter to anphibians
taxa_16s_raw = read.table(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_taxa_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_taxa_extracted"))), 
                    "/data/taxonomy.tsv")),
             header = TRUE, 
             sep = '\t',
             quote = '') %>%
  separate(Taxon, into = c("k", "p", "c", "o", "f", "g", "s"), sep = ";") %>%
  mutate(across(everything(), ~ str_remove(., "^[a-z]__"))) %>%
  filter(c == "Amphibia",
         s != "")

# rename columns
colnames(taxa_16s_raw) = c("feature_id",
                          "kingdom",
                          "phylum",
                          "class",
                          "order",
                          "family",
                          "genus",
                          "species")

# count by taxonomy
taxa_16s = taxa_16s_raw %>%
  group_by(order, family, genus, species) %>%
  summarise(n_16s = n(),
            .groups = "drop") %>%
  mutate(sn_16s = paste(genus, species))

# read in 12s sequences
seq_12s =
  read.fasta(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_seqs_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_seqs_extracted"))), 
                    "/data/dna-sequences.fasta")))

# read in 16s sequences
seq_16s =
  read.fasta(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_seqs_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_seqs_extracted"))), 
                    "/data/dna-sequences.fasta")))

```

# sequence counts
```{r}
# how many sequences per species?

## RIBBiTR amphibian species
taxa_rib_count = taxa_rib %>%
  left_join(taxa_12s %>%
              select(sn_12s,
                     n_12s), by = c("sn_rib" = "sn_12s")) %>%
  left_join(taxa_16s %>%
              select(sn_16s,
                     n_16s), by = c("sn_rib" = "sn_16s")) %>%
  pivot_longer(cols = c("n_12s", "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(n = if_else(is.na(n), 0, n),
         gene = gsub("n_", "", gene))

steps = seq(1, 50, 1)

# Combine into one table
cumulative_table_12s = tibble(
  sequences_n = steps,
  gene = "12s",
  n_species_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "12s" & taxa_rib_count$n == k)),
  n_species_gte_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "12s" & taxa_rib_count$n >= k)),
)

cumulative_table_16s = tibble(
  sequences_n = steps,
  gene = "16s",
  n_species_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "16s" & taxa_rib_count$n == k)),
  n_species_gte_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "16s" & taxa_rib_count$n >= k)),
)

cumulative_table = bind_rows(cumulative_table_12s,
                             cumulative_table_16s) %>%
  mutate(p_species_ii = n_species_ii / nrow(taxa_rib),
         p_species_gte_ii = n_species_gte_ii / nrow(taxa_rib))

# histogram of sequence counts
ggplot(taxa_rib_count, aes(x = n, fill = gene)) +
  geom_histogram(position = "identity", alpha = 0.5) +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "12S/16S sequence counts: RIBBiTR amphibian species",
       x = "Number of sequences (n)",
       y = "Number of species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

# proportion of RIBBiTR sequence counts
ggplot(cumulative_table, aes(x = sequences_n, y = p_species_gte_ii, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "Cumulative 12S/16S sequence counts: RIBBiTR amphibian species",
       x = "Number of sequences (n)",
       y = "Portion of RIBBiTR species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

## all amphibian species
taxa_total_count = taxa_12s %>%
  full_join(taxa_16s, by = c("order",
                             "family",
                             "genus",
                             "species",
                             "sn_12s" = "sn_16s")) %>%
  dplyr::rename(sn = sn_12s) %>%
  pivot_longer(cols = c("n_12s", "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(n = if_else(is.na(n), 0, n),
         gene = gsub("n_", "", gene))

cumulative_table_all = tibble(
  ii = steps,
  n_species_12s_gt_ii = sapply(steps, function(k) sum(taxa_total_count$gene == "12s" & taxa_total_count$n >= k)),
  n_species_16s_gt_ii = sapply(steps, function(k) sum(taxa_total_count$gene == "16s" & taxa_total_count$n >= k))
) %>%
  pivot_longer(cols = c("n_species_12s_gt_ii",
                        "n_species_16s_gt_ii")) %>%
  dplyr::rename(sequences_n = ii,
                species_gte_n = value) %>%
  mutate(gene = case_match(name,
                           "n_species_12s_gt_ii" ~ "12S",
                           "n_species_16s_gt_ii" ~ "16S",
                           .default = NA))


# ggplot(cumulative_table_all, aes(x = sequences_n, y = species_n, color = gene)) +
#   geom_line() +
#   scale_fill_brewer(palette = "Set2")

# histogram of sequence counts
ggplot(taxa_total_count, aes(x = n, fill = gene)) +
  geom_histogram(position = "identity", alpha = 0.5) +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "12S/16S sequence counts: All amphibian species",
       x = "Number of sequences (n)",
       y = "Number of species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

# proportion of RIBBiTR sequence counts
ggplot(cumulative_table_all, aes(x = sequences_n, y = species_gte_n, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "Cumulative 12S/16S sequence counts: All amphibian species",
       x = "Number of sequences (n)",
       y = "Number of species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

```

# similar species counts
```{r}
#| fig-height: 8
#| fig-width: 8

## RIBBiTR genera
cf_genus_rib = taxa_rib %>%
  group_by(itis_genus) %>%
  count(name = "n_ribbitr") %>%
  dplyr::rename(genus = "itis_genus")

cf_genus_12s = taxa_12s %>%
  select(genus, species) %>%
  distinct() %>%
  group_by(genus) %>%
  count(name = "n_12s")

cf_genus_16s = taxa_16s %>%
  select(genus, species) %>%
  distinct() %>%
  group_by(genus) %>%
  count(name = "n_16s")

cf_genus = cf_genus_rib %>%
  full_join(cf_genus_12s, by = "genus") %>%
  full_join(cf_genus_16s, by = "genus") %>%
  pivot_longer(cols = c("n_12s",
                        "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(gene = toupper(gsub("n_", "", gene)),
         n_ribbitr = if_else(is.na(n_ribbitr), 0, n_ribbitr),
         n = if_else(is.na(n), 0, n))

cf_genus_factor = cf_genus %>%
  filter(n_ribbitr > 0) %>%
  mutate(f = n / n_ribbitr)

genus_order = cf_genus_factor %>%
  filter(gene == "16S") %>%
  arrange(f) %>%
  pull(genus)

cf_genus_factor$genus = factor(cf_genus_factor$genus, levels = genus_order)

ggplot(cf_genus_factor, aes(x = f, y = genus, fill = gene)) +
  geom_col(position = position_dodge2(padding = 0)) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Amphibian 12S/16S sequence counts by RIBBiTR genera",
       x = "Ratio of Classifier species to RIBBiTR species within each genus",
       y = "Genus") +
  theme_minimal()

## RIBBiTR families
cf_family_rib = taxa_rib %>%
  group_by(itis_family) %>%
  count(name = "n_ribbitr") %>%
  dplyr::rename(family = "itis_family")

cf_family_12s = taxa_12s %>%
  select(family, genus, species) %>%
  distinct() %>%
  group_by(family) %>%
  count(name = "n_12s")

cf_family_16s = taxa_16s %>%
  select(family, genus, species) %>%
  distinct() %>%
  group_by(family) %>%
  count(name = "n_16s")

cf_family = cf_family_rib %>%
  full_join(cf_family_12s, by = "family") %>%
  full_join(cf_family_16s, by = "family") %>%
  pivot_longer(cols = c("n_12s",
                        "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(gene = toupper(gsub("n_", "", gene)),
         n_ribbitr = if_else(is.na(n_ribbitr), 0, n_ribbitr),
         n = if_else(is.na(n), 0, n))

cf_family_factor = cf_family %>%
  filter(n_ribbitr > 0) %>%
  mutate(f = n / n_ribbitr)

family_order = cf_family_factor %>%
  filter(gene == "16S") %>%
  arrange(f) %>%
  pull(family)

cf_family_factor$family = factor(cf_family_factor$family, levels = family_order)

ggplot(cf_family_factor, aes(x = f, y = family, fill = gene)) +
  geom_col(position = position_dodge2(padding = 0)) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Amphibian 12S/16S sequence counts by RIBBiTR family",
       x = "Ratio of Classifier species to RIBBiTR species within each family",
       y = "Family") +
  theme_minimal()

```

# sequence length distributions
```{r}
# look at sequence lengths (total and for RIBBiTR species)
taxa_seq_12s = taxa_12s_raw %>%
  mutate(sn = paste(genus, species)) %>%
  left_join(seq_12s, by = c("feature_id" = "seq.name")) %>%
  left_join(taxa_rib %>%
               select(taxon_id,
                      ncbi_taxon), by = c("sn" = "ncbi_taxon"))

taxa_seq_16s = taxa_16s_raw %>%
  mutate(sn = paste(genus, species)) %>%
  left_join(seq_16s, by = c("feature_id" = "seq.name")) %>%
  left_join(taxa_rib %>%
               select(taxon_id,
                      ncbi_taxon), by = c("sn" = "ncbi_taxon"))

# convert to DNAStringSets
set_12s = DNAStringSet(taxa_seq_12s$seq.text, use.name = TRUE)
names(set_12s) = taxa_seq_12s$feature_id

set_16s = DNAStringSet(taxa_seq_16s$seq.text, use.name = TRUE)
names(set_16s) = taxa_seq_16s$feature_id

width_df = data.frame(
  width = c(width(set_12s), width(set_16s)),
  gene = rep(c("12S", "16S"), c(length(set_12s), length(set_16s))),
  feature_id = c(names(set_12s), names(set_16s)),
  taxon_id = c(taxa_seq_12s$taxon_id, taxa_seq_16s$taxon_id)
)

ggplot(width_df, aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S sequence length distributions (all amphibians)",
       x = "Sequence length (bp)",
       y = "Sequences")

ggplot(width_df %>%
         filter(!is.na(taxon_id)), aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S sequence length distributions (RIBBiTR amphibians)",
       x = "Sequence length (bp)",
       y = "Sequences")
```

# within taxa similarity
```{r}

within_group_pid2 = function(df, group_col) {
  df %>%
    group_by(!!sym(group_col)) %>%
    reframe(
      n = n(),
      n2 = n ^ 2,
      {
        if (n > 1) {
          temp_mat <- 1 - as.matrix(DistanceMatrix(AlignSeqs(DNAStringSet(seq.text), verbose = FALSE), 
                                                   type = "dist", correction = NA, verbose = FALSE))
          tibble(pid = temp_mat[upper.tri(temp_mat)])
        } else {
          tibble(pid = NA)
        }
      }
    ) %>%
    ungroup()
}

spp_seq_12s = taxa_seq_12s %>%
  filter(sn %in% taxa_rib$sn_rib)

oput = within_group_pid2(spp_seq_12s, "species")

oput_samp = oput %>%
  filter(n > 1) %>%
  group_by(species) %>%
  slice_sample(n = 100) %>%
  ungroup()

# density plot
ggplot(oput_samp, aes(x = pid)) +
  geom_density()

steps = seq(1, 0.5, -0.01)

# Combine into one table
cumulative_oput = tibble(
  pid = steps,
  gene = "12s",
  pid_cum = sapply(steps, function(k) sum(oput_samp$pid >= k) / nrow(oput_samp)),
)

ggplot(cumulative_oput, aes(x = pid, y = pid_cum, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "Within-species identity (RIBBiTR species",
       x = "Within-species pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

# between taxa similarity
```{r}
pid_mat_f = function(seqs) {
  
  n = length(seqs)
  pid_mat <- matrix(0, n, n,
                    dimnames = list(names(seqs), names(seqs)))
  
  pb = txtProgressBar(min = 1, max = n, style = 3)
  for (ii in 1:n) {
    aln = pairwiseAlignment(seqs, seqs[[ii]])
    pid_mat[, ii] = pid(aln)
    
    setTxtProgressBar(pb, ii)
  }
  close(pb)
  
  return(pid_mat)
}

# 12s
genus_seq_12s = taxa_seq_12s %>%
  filter(genus %in% taxa_rib$ncbi_genus)

consensus_genus_12s = genus_seq_12s %>%
  group_by(genus) %>%
  summarise(
    consensus = as.character(consensusString(DNAStringSet(seq.text))),
    .groups = "drop"
  )

consensus_genus_12s_set = DNAStringSet(consensus_genus_12s$consensus, use.name = TRUE)
names(consensus_genus_12s_set) = consensus_genus_12s$genus

between_genus_pid_12s = pid_mat_f(consensus_genus_12s_set)

bg_pid_12s = between_genus_pid_12s[upper.tri(between_genus_pid_12s)]


# between_genus_dist_12s = stringDist(consensus_genus_12s_set, method = "levenshtein")
# 
# between_genus_12s = as.data.frame(as.matrix(between_genus_dist_12s)) %>%
#   rownames_to_column("genus1") %>%
#   pivot_longer(-genus1, names_to = "genus2", values_to = "distance") %>%
#   filter(genus1 < genus2) %>%
#   mutate(gene = "12s")

## 16s
genus_seq_16s = taxa_seq_16s %>%
  filter(genus %in% taxa_rib$ncbi_genus)

upper_tri_vec <- pid_matrix[upper.tri(pid_matrix)]consensus_genus_16s = genus_seq_16s %>%
  group_by(genus) %>%
  summarise(
    consensus = as.character(consensusString(DNAStringSet(seq.text))),
    .groups = "drop"
  )

consensus_genus_16s_set = DNAStringSet(consensus_genus_16s$consensus, use.name = TRUE)
names(consensus_genus_16s_set) = consensus_genus_16s$genus

between_genus_pid_16s = pid_mat_f(consensus_genus_16s_set)

bg_pid_16s = between_genus_pid_16s[upper.tri(between_genus_pid_16s)]

between_genus_pid = data.frame(
  pid = c(bg_pid_12s, bg_pid_16s),
  gene = rep(c("12S", "16S"), c(length(bg_pid_12s), length(bg_pid_16s)))
)

ggplot(between_genus_pid, aes(x = pid, fill = gene)) +
  geom_density(alpha = 0.6) +
  theme_minimal() +
  xlim(0, 100)

```
