---
title: "0b_species_list_compare"
output: html_document
---

# setup
```{r}
# setup

# minimal libraries
librarian::shelf(tidyverse, dbplyr, RPostgres, DBI, RIBBiTR-BII/ribbitrrr, yaml, here, phylotools, Biostrings, pwalign, DECIPHER, nls2)

# read in config files
config_12s = read_yaml(here("16S Process TEST", "runs/methods_2026-01-12/12S/output/metadata/config.yml"))
config_16s = read_yaml(here("16S Process TEST", "runs/methods_2026-01-12/16S/output/metadata/config.yml"))

# establish database connection
dbcon = hopToDB("ribbitr")

# download RIBBiTR species list, filter to amphibians
taxa_rib = tbl(dbcon, Id("survey_data", "taxonomy")) %>%
  filter(itis_rank == "species",
         itis_class == "Amphibia") %>%
  select(taxon_id,
         itis_tsn,
         itis_order,
         itis_family,
         itis_genus,
         itis_species,
         ncbi_id,
         ncbi_taxon) %>%
  collect() %>%
  mutate(sn_rib = coalesce(ncbi_taxon, itis_species)) %>%
  separate(sn_rib, c("ncbi_genus", "ncbi_species"), sep = " ", remove = FALSE)

# check for discrepancies between itis and ncbi
itis_ncbi_dif = taxa_rib %>%
  filter(itis_species != ncbi_taxon)

# read in 12s taxonomies, filter to amphibians
taxa_12s_raw = read.table(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_taxa_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_taxa_extracted"))), 
                    "/data/taxonomy.tsv")),
             header = TRUE, 
             sep = '\t',
             quote = '') %>%
  separate(Taxon, into = c("k", "p", "c", "o", "f", "g", "s"), sep = ";") %>%
  mutate(across(everything(), ~ str_remove(., "^[a-z]__"))) %>%
  filter(c == "Amphibia",
         s != "")

# rename columns
colnames(taxa_12s_raw) = c("feature_id",
                          "kingdom",
                          "phylum",
                          "class",
                          "order",
                          "family",
                          "genus",
                          "species")

# count by taxonomy
taxa_12s = taxa_12s_raw %>%
  group_by(order, family, genus, species) %>%
  summarise(n_12s = n(),
            .groups = "drop") %>%
  mutate(sn_12s = paste(genus, species))

# read in 16s taxonomy, filter to anphibians
taxa_16s_raw = read.table(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_taxa_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_taxa_extracted"))), 
                    "/data/taxonomy.tsv")),
             header = TRUE, 
             sep = '\t',
             quote = '') %>%
  separate(Taxon, into = c("k", "p", "c", "o", "f", "g", "s"), sep = ";") %>%
  mutate(across(everything(), ~ str_remove(., "^[a-z]__"))) %>%
  filter(c == "Amphibia",
         s != "")

# rename columns
colnames(taxa_16s_raw) = c("feature_id",
                          "kingdom",
                          "phylum",
                          "class",
                          "order",
                          "family",
                          "genus",
                          "species")

# count by taxonomy
taxa_16s = taxa_16s_raw %>%
  group_by(order, family, genus, species) %>%
  summarise(n_16s = n(),
            .groups = "drop") %>%
  mutate(sn_16s = paste(genus, species))

# read in 12s classifier sequences
seq_12s =
  read.fasta(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_seqs_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_12s$taxonomy$classifierDir, "/Vertebrata", config_12s$taxonomy$gene, "_derep1_seqs_extracted"))), 
                    "/data/dna-sequences.fasta")))

# read in 16s classifier sequences
seq_16s =
  read.fasta(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_seqs_extracted/", 
                    list.files(here("16S Process TEST", paste0(config_16s$taxonomy$classifierDir, "/Vertebrata", config_16s$taxonomy$gene, "_derep1_seqs_extracted"))), 
                    "/data/dna-sequences.fasta")))

```

# samples
```{r}
# 12s
feature_table_12s =
  read_tsv(here("16S Process TEST", config_12s$run$runDir, "analysis/s06_denoised_12S_eDNA/sample_table/feature-table.tsv"), skip = 1) %>%
  as.data.frame() %>%
  dplyr::rename("asv_id" = "#OTU ID") %>%
  arrange(asv_id)


all_ids_12s = feature_table_12s %>%
  pivot_longer(cols = -asv_id,
               names_to = "illumina_sample_id",
               values_to = "asv_count") %>%
  select(illumina_sample_id) %>%
  distinct() %>%
  mutate(crispr = grepl("CRISPR", illumina_sample_id),
         block = FALSE,
         control = grepl("Negative|Positive", illumina_sample_id)) %>%
  arrange(illumina_sample_id)

feature_long_12s = feature_table_12s %>%
  pivot_longer(cols = -asv_id,
               names_to = "illumina_sample_id",
               values_to = "asv_count") %>%
  filter(asv_count != 0) %>%
  select(illumina_sample_id,
         everything()) %>%
  arrange(illumina_sample_id,
          desc(asv_count))

# 16s
feature_table_16s =
  read_tsv(here("16S Process TEST", config_16s$run$runDir, "analysis/s06_denoised_16S_eDNA/sample_table/feature-table.tsv"), skip = 1) %>%
  as.data.frame() %>%
  dplyr::rename("asv_id" = "#OTU ID") %>%
  arrange(asv_id)


all_ids_16s = feature_table_16s %>%
  pivot_longer(cols = -asv_id,
               names_to = "illumina_sample_id",
               values_to = "asv_count") %>%
  select(illumina_sample_id) %>%
  distinct() %>%
  mutate(crispr = grepl("CRISPR", illumina_sample_id),
         block = grepl("block", illumina_sample_id) & !grepl("noblock", illumina_sample_id),
         control = grepl("Negative|Positive|Neg|BH", illumina_sample_id)) %>%
  arrange(illumina_sample_id)

feature_long_16s = feature_table_16s %>%
  pivot_longer(cols = -asv_id,
               names_to = "illumina_sample_id",
               values_to = "asv_count") %>%
  filter(asv_count != 0) %>%
  select(illumina_sample_id,
         everything()) %>%
  arrange(illumina_sample_id,
          desc(asv_count))

```
# asv sequences
```{r}
asv_seq_12s =
  read.fasta(here("16S Process TEST", paste0(config_12s$run$runDir, "/analysis/s06_denoised_", config_12s$taxonomy$gene, "_eDNA/representative_sequences/", 
                           list.files(here("16S Process TEST", paste0(config_12s$run$runDir, "/analysis/s06_denoised_", config_12s$taxonomy$gene, "_eDNA/representative_sequences"))), 
                           "/data/dna-sequences.fasta"))) %>%
  dplyr::rename(asv_id = seq.name) %>%
  arrange(asv_id)

asv_seq_16s =
  read.fasta(here("16S Process TEST", paste0(config_16s$run$runDir, "/analysis/s06_denoised_", config_16s$taxonomy$gene, "_eDNA/representative_sequences/", 
                           list.files(here("16S Process TEST", paste0(config_16s$run$runDir, "/analysis/s06_denoised_", config_16s$taxonomy$gene, "_eDNA/representative_sequences"))), 
                           "/data/dna-sequences.fasta"))) %>%
  dplyr::rename(asv_id = seq.name) %>%
  arrange(asv_id)

```

# hits
```{r}
hits_12s = read_csv(here("16S Process TEST", config_12s$run$runDir, "output", paste0(config_12s$run$name, "_eDNA_hits_joined.csv"))) %>%
  filter(!is.na(class),
         method == "vsearch",
         seq_overlap >= 50) %>%
  select(-consensus) %>%
  mutate(p_identical = percent_identical / 100,
         p_overlap = seq_overlap / (s_end - s_start + 1))

hits_16s = read_csv(here("16S Process TEST", config_16s$run$runDir, "output", paste0(config_16s$run$name, "_eDNA_hits_joined.csv"))) %>%
  filter(!is.na(class),
         method == "vsearch",
         seq_overlap >= 50) %>%
  select(-consensus) %>%
  mutate(p_identical = percent_identical / 100,
         p_overlap = seq_overlap / (s_end - s_start + 1))


```

# sequence counts
```{r}
# how many sequences per species?

## RIBBiTR amphibian species
taxa_rib_count = taxa_rib %>%
  left_join(taxa_12s %>%
              select(sn_12s,
                     n_12s), by = c("sn_rib" = "sn_12s")) %>%
  left_join(taxa_16s %>%
              select(sn_16s,
                     n_16s), by = c("sn_rib" = "sn_16s")) %>%
  pivot_longer(cols = c("n_12s", "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(n = if_else(is.na(n), 0, n),
         gene = gsub("n_", "", gene))

steps = seq(1, 50, 1)

# Combine into one table
cumulative_table_12s = tibble(
  sequences_n = steps,
  gene = "12s",
  n_species_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "12s" & taxa_rib_count$n == k)),
  n_species_gte_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "12s" & taxa_rib_count$n >= k)),
)

cumulative_table_16s = tibble(
  sequences_n = steps,
  gene = "16s",
  n_species_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "16s" & taxa_rib_count$n == k)),
  n_species_gte_ii = sapply(steps, function(k) sum(taxa_rib_count$gene == "16s" & taxa_rib_count$n >= k)),
)

cumulative_table = bind_rows(cumulative_table_12s,
                             cumulative_table_16s) %>%
  mutate(p_species_ii = n_species_ii / nrow(taxa_rib),
         p_species_gte_ii = n_species_gte_ii / nrow(taxa_rib))

# histogram of sequence counts
ggplot(taxa_rib_count, aes(x = n, fill = gene)) +
  geom_histogram(position = "identity", alpha = 0.5) +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "12S/16S sequence counts: RIBBiTR amphibian species",
       x = "Number of sequences (n)",
       y = "Number of species with 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

# proportion of RIBBiTR sequence counts
ggplot(cumulative_table, aes(x = sequences_n, y = p_species_gte_ii, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "Cumulative 12S/16S sequence counts: RIBBiTR amphibian species",
       x = "Minimum number of sequences (n)",
       y = "Portion of RIBBiTR species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

## all amphibian species
taxa_total_count = taxa_12s %>%
  full_join(taxa_16s, by = c("order",
                             "family",
                             "genus",
                             "species",
                             "sn_12s" = "sn_16s")) %>%
  dplyr::rename(sn = sn_12s) %>%
  pivot_longer(cols = c("n_12s", "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(n = if_else(is.na(n), 0, n),
         gene = gsub("n_", "", gene))

cumulative_table_all = tibble(
  ii = steps,
  n_species_12s_gt_ii = sapply(steps, function(k) sum(taxa_total_count$gene == "12s" & taxa_total_count$n >= k)),
  n_species_16s_gt_ii = sapply(steps, function(k) sum(taxa_total_count$gene == "16s" & taxa_total_count$n >= k))
) %>%
  pivot_longer(cols = c("n_species_12s_gt_ii",
                        "n_species_16s_gt_ii")) %>%
  dplyr::rename(sequences_n = ii,
                species_gte_n = value) %>%
  mutate(gene = recode_values(name,
                           "n_species_12s_gt_ii" ~ "12S",
                           "n_species_16s_gt_ii" ~ "16S",
                           default = NA))


# ggplot(cumulative_table_all, aes(x = sequences_n, y = species_n, color = gene)) +
#   geom_line() +
#   scale_fill_brewer(palette = "Set2")

# histogram of sequence counts
ggplot(taxa_total_count, aes(x = n, fill = gene)) +
  geom_histogram(position = "identity", alpha = 0.5) +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "12S/16S sequence counts: All amphibian species",
       x = "Number of sequences (n)",
       y = "Number of species with 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

# proportion of RIBBiTR sequence counts
ggplot(cumulative_table_all, aes(x = sequences_n, y = species_gte_n, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2")+
  labs(title = "Cumulative 12S/16S sequence counts: All amphibian species",
       x = "Minimum number of sequences (n)",
       y = "Number of species with at least 'n' sequences") +
  xlim(1, 50) +
  theme_minimal()

```

# similar species counts
```{r}
#| fig-height: 8
#| fig-width: 8

## RIBBiTR genera
cf_genus_rib = taxa_rib %>%
  group_by(itis_genus) %>%
  count(name = "n_ribbitr") %>%
  dplyr::rename(genus = "itis_genus")

cf_genus_12s = taxa_12s %>%
  select(genus, species) %>%
  distinct() %>%
  group_by(genus) %>%
  count(name = "n_12s")

cf_genus_16s = taxa_16s %>%
  select(genus, species) %>%
  distinct() %>%
  group_by(genus) %>%
  count(name = "n_16s")

cf_genus = cf_genus_rib %>%
  full_join(cf_genus_12s, by = "genus") %>%
  full_join(cf_genus_16s, by = "genus") %>%
  pivot_longer(cols = c("n_12s",
                        "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(gene = toupper(gsub("n_", "", gene)),
         n_ribbitr = if_else(is.na(n_ribbitr), 0, n_ribbitr),
         n = if_else(is.na(n), 0, n))

cf_genus_factor = cf_genus %>%
  filter(n_ribbitr > 0) %>%
  mutate(f = n / n_ribbitr)

genus_order = cf_genus_factor %>%
  filter(gene == "16S") %>%
  arrange(f) %>%
  pull(genus)

cf_genus_factor$genus = factor(cf_genus_factor$genus, levels = genus_order)

ggplot(cf_genus_factor, aes(x = f, y = genus, fill = gene)) +
  geom_col(position = position_dodge2(padding = 0)) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Amphibian 12S/16S sequence counts by RIBBiTR genera",
       x = "Ratio of Classifier species to RIBBiTR species within each genus",
       y = "Genus") +
  theme_minimal()

## RIBBiTR families
cf_family_rib = taxa_rib %>%
  group_by(itis_family) %>%
  count(name = "n_ribbitr") %>%
  dplyr::rename(family = "itis_family")

cf_family_12s = taxa_12s %>%
  select(family, genus, species) %>%
  distinct() %>%
  group_by(family) %>%
  count(name = "n_12s")

cf_family_16s = taxa_16s %>%
  select(family, genus, species) %>%
  distinct() %>%
  group_by(family) %>%
  count(name = "n_16s")

cf_family = cf_family_rib %>%
  full_join(cf_family_12s, by = "family") %>%
  full_join(cf_family_16s, by = "family") %>%
  pivot_longer(cols = c("n_12s",
                        "n_16s"),
               names_to = "gene",
               values_to = "n") %>%
  mutate(gene = toupper(gsub("n_", "", gene)),
         n_ribbitr = if_else(is.na(n_ribbitr), 0, n_ribbitr),
         n = if_else(is.na(n), 0, n))

cf_family_factor = cf_family %>%
  filter(n_ribbitr > 0) %>%
  mutate(f = n / n_ribbitr)

family_order = cf_family_factor %>%
  filter(gene == "16S") %>%
  arrange(f) %>%
  pull(family)

cf_family_factor$family = factor(cf_family_factor$family, levels = family_order)

ggplot(cf_family_factor, aes(x = f, y = family, fill = gene)) +
  geom_col(position = position_dodge2(padding = 0)) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Amphibian 12S/16S sequence counts by RIBBiTR family",
       x = "Ratio of Classifier species to RIBBiTR species within each family",
       y = "Family") +
  theme_minimal()

```

# sequence length distributions: classifiers
```{r}
# look at sequence lengths (total and for RIBBiTR species)
taxa_seq_12s = taxa_12s_raw %>%
  mutate(sn = paste(genus, species)) %>%
  left_join(seq_12s, by = c("feature_id" = "seq.name")) %>%
  left_join(taxa_rib %>%
               select(taxon_id,
                      ncbi_taxon), by = c("sn" = "ncbi_taxon"))

taxa_seq_16s = taxa_16s_raw %>%
  mutate(sn = paste(genus, species)) %>%
  left_join(seq_16s, by = c("feature_id" = "seq.name")) %>%
  left_join(taxa_rib %>%
               select(taxon_id,
                      ncbi_taxon), by = c("sn" = "ncbi_taxon"))

# convert to DNAStringSets
set_12s = DNAStringSet(taxa_seq_12s$seq.text, use.name = TRUE)
names(set_12s) = taxa_seq_12s$feature_id

set_16s = DNAStringSet(taxa_seq_16s$seq.text, use.name = TRUE)
names(set_16s) = taxa_seq_16s$feature_id

width_df = data.frame(
  width = c(width(set_12s), width(set_16s)),
  gene = rep(c("12S", "16S"), c(length(set_12s), length(set_16s))),
  feature_id = c(names(set_12s), names(set_16s)),
  taxon_id = c(taxa_seq_12s$taxon_id, taxa_seq_16s$taxon_id)
)

# all taxa in classifier (vertebrata)
# convert to DNAStringSets
set_12s_all = DNAStringSet(seq_12s$seq.text, use.name = TRUE)

set_16s_all = DNAStringSet(seq_16s$seq.text, use.name = TRUE)

width_df_all = data.frame(
  width = c(width(set_12s_all), width(set_16s_all)),
  gene = rep(c("12S", "16S"), c(length(set_12s_all), length(set_16s_all)))
)

ggplot(width_df_all, aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S sequence length distributions (Vertebrata)",
       x = "Sequence length (bp)",
       y = "Sequences")

ggplot(width_df, aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S sequence length distributions (Amphibia)",
       x = "Sequence length (bp)",
       y = "Sequences")

ggplot(width_df %>%
         filter(!is.na(taxon_id)), aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S sequence length distributions (RIBBiTR amphibians)",
       x = "Sequence length (bp)",
       y = "Sequences")
```
# sequence length distributions: ASVs
```{r}
# convert to DNAStringSets
set_asv_12s = DNAStringSet(asv_seq_12s$seq.text, use.name = TRUE)
names(set_asv_12s) = asv_seq_12s$asv_id

set_asv_16s = DNAStringSet(asv_seq_16s$seq.text, use.name = TRUE)
names(set_asv_16s) = asv_seq_16s$asv_id

width_asv_df = data.frame(
  width = c(width(set_asv_12s), width(set_asv_16s)),
  gene = rep(c("12S", "16S"), c(length(set_asv_12s), length(set_asv_16s))),
  feature_id = c(names(set_asv_12s), names(set_asv_16s))
)

ggplot(width_asv_df, aes(x = width, fill = gene)) +
  geom_histogram() +
  facet_wrap("gene", scales = "free") +
  theme_minimal() +
  labs(title = "12S/16S ASV sequence length distributions",
       x = "Sequence length (bp)",
       y = "Sequences")
```
# Taxa similarity
```{r}

within_group_pid = function(df, group_col) {
  df %>%
    group_by(!!sym(group_col)) %>%
    reframe(
      n = n(),
      n2 = n ^ 2,
      {
        if (n > 1) {
          temp_mat <- 1 - as.matrix(DistanceMatrix(AlignSeqs(DNAStringSet(seq.text), verbose = FALSE), 
                                                   type = "matrix", correction = NA, verbose = FALSE))
          tibble(pid = temp_mat[upper.tri(temp_mat)])
        } else {
          tibble(pid = NA)
        }
      }
    ) %>%
    ungroup()
}

df = peace
group_col = "genus"

between_group_pid = function(df, group_col, replicate = 10) {
  testme = df %>%
    group_by(!!sym(group_col)) %>%
    slice_sample(n = replicate, replace = TRUE) %>%
    mutate(sample_group = row_number()) %>%
    ungroup() %>%
    group_by(sample_group) %>%
    reframe(
      n = n(),
      n2 = n ^ 2,
      {
        if (n > 1) {
          temp_mat <- 1 - as.matrix(DistanceMatrix(AlignSeqs(DNAStringSet(seq.text), verbose = FALSE), 
                                                   type = "matrix", correction = NA, verbose = FALSE))
          tibble(pid = temp_mat[upper.tri(temp_mat)])
        } else {
          tibble(pid = NA)
        }
      }
    ) %>%
    ungroup()
}

cumulative_gte = function(values, start, end, step){
  steps = seq(start, end, step)
  
  cumulative_oput = tibble(
    x = steps,
    val_cum = sapply(steps, function(k) sum(values >= k) / length(values)),
  )

}
```

## 12S

### 12S species
```{r}
species_within_pid_12s = taxa_seq_12s %>%
  filter(sn %in% taxa_rib$sn_rib) %>%
  within_group_pid("species")

species_within_pid_12s_samp = species_within_pid_12s %>%
  filter(n > 1) %>%
  group_by(species) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "12s",
         type = "within",
         level = "species")

species_within_pid_12s_cum = cumulative_gte(species_within_pid_12s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "within",
         level = "species")

species_between_pid_12s = taxa_seq_12s %>%
  filter(sn %in% taxa_rib$sn_rib) %>%
  between_group_pid("species", replicate = 10) %>%
  mutate(gene = "12s",
         type = "between",
         level = "species")

species_between_pid_12s = taxa_seq_12s %>%
  
  # filter(sn %in% taxa_rib$sn_rib) %>%
  between_group_pid("species", replicate = 10) %>%
  mutate(gene = "12s",
         type = "between",
         level = "species")

species_between_pid_12s_cum = cumulative_gte(species_between_pid_12s$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "between",
         level = "species")

species_pid_12s = bind_rows(species_within_pid_12s_samp,
                            species_between_pid_12s)

species_pid_12s_cum = bind_rows(species_within_pid_12s_cum,
                            species_between_pid_12s_cum)

# calculate numbers of sequences

## median number of sequences in species
n_in = 

## median number of sequences out of species


# density plot
ggplot(species_pid_12s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(species_pid_12s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

### 12S genus
```{r}
genus_within_pid_12s = taxa_seq_12s %>%
  filter(genus %in% taxa_rib$ncbi_genus) %>%
  within_group_pid("genus")

genus_within_pid_12s_samp = genus_within_pid_12s %>%
  filter(n > 1) %>%
  group_by(genus) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "12s",
         type = "within",
         level = "genus")

genus_within_pid_12s_cum = cumulative_gte(genus_within_pid_12s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "within",
         level = "genus")

genus_between_pid_12s = taxa_seq_12s %>%
  filter(genus %in% taxa_rib$ncbi_genus) %>%
  between_group_pid("genus", replicate = 10) %>%
  mutate(gene = "12s",
         type = "between",
         level = "genus")

genus_between_pid_12s_cum = cumulative_gte(genus_between_pid_12s$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "between",
         level = "genus")

genus_pid_12s = bind_rows(genus_within_pid_12s_samp,
                            genus_between_pid_12s)

genus_pid_12s_cum = bind_rows(genus_within_pid_12s_cum,
                            genus_between_pid_12s_cum)

# density plot
ggplot(genus_pid_12s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(genus_pid_12s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

### 12S family
```{r}
family_within_pid_12s = taxa_seq_12s %>%
  filter(family %in% taxa_rib$itis_family) %>%
  within_group_pid("family")

family_within_pid_12s_samp = family_within_pid_12s %>%
  filter(n > 1) %>%
  group_by(family) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "12s",
         type = "within")

family_within_pid_12s_cum = cumulative_gte(family_within_pid_12s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "within",
         level = "family")

family_between_pid_12s = taxa_seq_12s %>%
  filter(family %in% taxa_rib$itis_family) %>%
  between_group_pid("family", replicate = 10) %>%
  mutate(gene = "12s",
         type = "between",
         level = "family")

family_between_pid_12s_cum = cumulative_gte(family_between_pid_12s$pid, 1, 0, -0.02) %>%
  mutate(gene = "12s",
         type = "between",
         level = "family")

family_pid_12s = bind_rows(family_within_pid_12s_samp,
                            family_between_pid_12s)

family_pid_12s_cum = bind_rows(family_within_pid_12s_cum,
                            family_between_pid_12s_cum)

# density plot
ggplot(family_pid_12s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(family_pid_12s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

## 16S
### 16S species
```{r}
species_within_pid_16s = taxa_seq_16s %>%
  filter(sn %in% taxa_rib$sn_rib) %>%
  within_group_pid("species")

species_within_pid_16s_samp = species_within_pid_16s %>%
  filter(n > 1) %>%
  group_by(species) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "16s",
         type = "within",
         level = "species")

species_within_pid_16s_cum = cumulative_gte(species_within_pid_16s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "within",
         level = "species")

species_between_pid_16s = taxa_seq_16s %>%
  filter(sn %in% taxa_rib$sn_rib) %>%
  between_group_pid("species", replicate = 10) %>%
  mutate(gene = "16s",
         type = "between",
         level = "species")

species_between_pid_16s_cum = cumulative_gte(species_between_pid_16s$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "between",
         level = "species")

species_pid_16s = bind_rows(species_within_pid_16s_samp,
                            species_between_pid_16s)

species_pid_16s_cum = bind_rows(species_within_pid_16s_cum,
                            species_between_pid_16s_cum)

# density plot
ggplot(species_pid_16s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(species_pid_16s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

### 16s genus
```{r}
genus_within_pid_16s = taxa_seq_16s %>%
  filter(genus %in% taxa_rib$ncbi_genus) %>%
  within_group_pid("genus")

genus_within_pid_16s_samp = genus_within_pid_16s %>%
  filter(n > 1) %>%
  group_by(genus) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "16s",
         type = "within",
         level = "genus")

genus_within_pid_16s_cum = cumulative_gte(genus_within_pid_16s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "within",
         level = "genus")

genus_between_pid_16s = taxa_seq_16s %>%
  filter(genus %in% taxa_rib$ncbi_genus) %>%
  between_group_pid("genus", replicate = 10) %>%
  mutate(gene = "16s",
         type = "between",
         level = "genus")

genus_between_pid_16s_cum = cumulative_gte(genus_between_pid_16s$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "between",
         level = "genus")

genus_pid_16s = bind_rows(genus_within_pid_16s_samp,
                            genus_between_pid_16s)

genus_pid_16s_cum = bind_rows(genus_within_pid_16s_cum,
                            genus_between_pid_16s_cum)

# density plot
ggplot(genus_pid_16s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(genus_pid_16s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

### 16s family
```{r}
family_within_pid_16s = taxa_seq_16s %>%
  filter(family %in% taxa_rib$itis_family) %>%
  within_group_pid("family")

family_within_pid_16s_samp = family_within_pid_16s %>%
  filter(n > 1) %>%
  group_by(family) %>%
  slice_sample(n = 100, replace = TRUE) %>%
  ungroup() %>%
  mutate(gene = "16s",
         type = "within")

family_within_pid_16s_cum = cumulative_gte(family_within_pid_16s_samp$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "within",
         level = "family")

family_between_pid_16s = taxa_seq_16s %>%
  filter(family %in% taxa_rib$itis_family) %>%
  between_group_pid("family", replicate = 10) %>%
  mutate(gene = "16s",
         type = "between",
         level = "family")

family_between_pid_16s_cum = cumulative_gte(family_between_pid_16s$pid, 1, 0, -0.02) %>%
  mutate(gene = "16s",
         type = "between",
         level = "family")

family_pid_16s = bind_rows(family_within_pid_16s_samp,
                            family_between_pid_16s)

family_pid_16s_cum = bind_rows(family_within_pid_16s_cum,
                            family_between_pid_16s_cum)

# density plot
ggplot(family_pid_16s, aes(x = pid, color = type)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal()

# cumulative plot
ggplot(family_pid_16s_cum, aes(x = x, y = val_cum, color = type)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal()


```

## Combined
```{r}

species_pid = bind_rows(species_pid_12s,
                       species_pid_16s)

species_pid_cum = bind_rows(species_pid_12s_cum,
                           species_pid_16s_cum)

# density plot
ggplot(species_pid, aes(x = pid, color = gene)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal() +
  facet_wrap("type")

# cumulative plot
ggplot(species_pid_cum, aes(x = x, y = val_cum, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-species identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal() +
  facet_wrap("type")

# genus
genus_pid = bind_rows(genus_pid_12s,
                       genus_pid_16s)

genus_pid_cum = bind_rows(genus_pid_12s_cum,
                           genus_pid_16s_cum)

# density plot
ggplot(genus_pid, aes(x = pid, color = gene)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal() +
  facet_wrap("type")

# cumulative plot
ggplot(genus_pid_cum, aes(x = x, y = val_cum, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-genus identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal() +
  facet_wrap("type")


# family
family_pid = bind_rows(family_pid_12s,
                       family_pid_16s)

family_pid_cum = bind_rows(family_pid_12s_cum,
                           family_pid_16s_cum)

# density plot
ggplot(family_pid, aes(x = pid, color = gene)) +
  geom_density() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "Relative density") +
  theme_minimal() +
  facet_wrap("type")

# cumulative plot
ggplot(family_pid_cum, aes(x = x, y = val_cum, color = gene)) +
  geom_line() +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Within- & between-family identity (RIBBiTR taxa)",
       x = "Pairwise fractional identity",
       y = "1 - Cumulative density") +
  theme_minimal() +
  facet_wrap("type")


```

# ASVs per sample
```{r}
asv_rarefy <- function(df, max_n, ii = 100) {
  samples <- unique(df$illumina_sample_id)
  n_samples <- length(samples)
  if (max_n > n_samples) {
    warning("max_n exceeds number of samples; setting to n_samples")
    max_n <- n_samples
  }
  
  nn_vec <- 0:max_n
  
  results <- map_dfr(nn_vec, function(nn) {
    sims <- map_df(1:ii, ~ {
      sel_samples <- sample(samples, size = nn, replace = FALSE)
      df %>% 
        filter(illumina_sample_id %in% sel_samples) %>% 
        summarise(n_asv = n_distinct(asv_id), .groups = "drop")
    })
    sims %>% 
      summarise(
        nn = nn,
        median_asv = median(n_asv),
        q25 = quantile(n_asv, 0.25),
        q75 = quantile(n_asv, 0.75),
        mean_asv = mean(n_asv),
        sd_asv = sd(n_asv),
        .groups = "drop"
      )
  })
  
  return(results)
}

# 12s
## filter to simple samples
simple_ids_12s = all_ids_12s %>%
  filter(!crispr,
         !block,
         !control)

simple_asvs_12s = feature_long_12s %>%
  filter(illumina_sample_id %in% simple_ids_12s$illumina_sample_id)

mean_asv_per_samp_12s = nrow(simple_asvs_12s) / nrow(simple_ids_12s)

rarefied_12s = asv_rarefy(simple_asvs_12s, nrow(simple_ids_12s), ii = 100) %>%
  mutate(gene = "12S",
         mean_asv_0 = nn * mean_asv_per_samp_12s)

# 16s
## filter to simple samples
simple_ids_16s = all_ids_16s %>%
  filter(!crispr,
         !block,
         !control)

simple_asvs_16s = feature_long_16s %>%
  filter(illumina_sample_id %in% simple_ids_16s$illumina_sample_id)

mean_asv_per_samp_16s = nrow(simple_asvs_16s) / nrow(simple_ids_16s)

rarefied_16s = asv_rarefy(df = simple_asvs_16s, max_n = nrow(simple_ids_16s), ii = 100) %>%
  mutate(gene = "16S",
         mean_asv_0 = nn * mean_asv_per_samp_16s)

rarefied = bind_rows(rarefied_12s,
                     rarefied_16s) %>%
  mutate(mean_asv_per_sample = mean_asv / nn)

ggplot(rarefied, aes(x = nn, y = median_asv, color = gene, fill = gene)) +
  geom_line() +
  geom_line(aes(y = mean_asv_0), linetype = 2) +
  ylim(0, 200)

ggplot(rarefied, aes(x = nn, y = mean_asv_per_sample, color = gene, fill = gene)) +
  geom_line() +
  ylim(0, NA)

# estimate asymptote

# Inverse: 1/S vs 1/nn should be linear
lin_data_12s <- rarefied_12s %>% 
  filter(nn > 5, median_asv > 0) %>%  # Avoid edge effects
  mutate(inv_S = 1/median_asv, inv_nn = 1/nn)

lm_fit_12s <- lm(inv_S ~ inv_nn, data = lin_data_12s)
asymptote_lin_12s <- 1/coef(lm_fit_12s)[1]  # Intercept gives 1/K
asymptote_lin_12s

lin_data_16s <- rarefied_16s %>% 
  filter(nn > 5, median_asv > 0) %>%  # Avoid edge effects
  mutate(inv_S = 1/median_asv, inv_nn = 1/nn)

lm_fit_16s <- lm(inv_S ~ inv_nn, data = lin_data_16s)
asymptote_lin_16s <- 1/coef(lm_fit_16s)[1]  # Intercept gives 1/K
asymptote_lin_16s
```

# Hits per sample
```{r}
# 12s
## filter to simple samples
simple_hits_12s = hits_12s %>%
  select(asv_id,
         taxon_id,
         class,
         percent_identical,
         seq_overlap) %>%
  inner_join(simple_asvs_12s, by = "asv_id", relationship = "many-to-many")

quality_hits_12s = simple_hits_12s %>%
  filter(percent_identical >= 95)

simple_hits_amphibia_12s = simple_hits_12s %>%
  filter(class == "Amphibia")

quality_hits_amphibia_12s = quality_hits_12s %>%
  filter(class == "Amphibia")

mean_hit_per_asv_12s = nrow(simple_hits_12s) / nrow(simple_asvs_12s)
mean_amphibia_hit_per_asv_12s = nrow(simple_hits_amphibia_12s) / nrow(simple_asvs_12s)

any_hit_per_asv_12s = nrow(simple_hits_12s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_12s)
any_amphibia_hit_per_asv_12s = nrow(simple_hits_amphibia_12s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_12s)

quality_hit_per_asv_12s = nrow(quality_hits_12s) / nrow(simple_asvs_12s)
quality_amphibia_hit_per_asv_12s = nrow(quality_hits_amphibia_12s) / nrow(simple_asvs_12s)

any_quality_hit_per_asv_12s = nrow(quality_hits_12s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_12s)
any_quality_amphibia_hit_per_asv_12s = nrow(quality_hits_amphibia_12s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_12s)

# 16s
## filter to simple samples
simple_hits_16s = hits_16s %>%
  select(asv_id,
         taxon_id,
         class,
         percent_identical,
         seq_overlap) %>%
  inner_join(simple_asvs_16s, by = "asv_id", relationship = "many-to-many")

quality_hits_16s = simple_hits_16s %>%
  filter(percent_identical >= 95)

simple_hits_amphibia_16s = simple_hits_16s %>%
  filter(class == "Amphibia")

quality_hits_amphibia_16s = quality_hits_16s %>%
  filter(class == "Amphibia")

mean_hit_per_asv_16s = nrow(simple_hits_16s) / nrow(simple_asvs_16s)
mean_amphibia_hit_per_asv_16s = nrow(simple_hits_amphibia_16s) / nrow(simple_asvs_16s)

any_hit_per_asv_16s = nrow(simple_hits_16s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_16s)
any_amphibia_hit_per_asv_16s = nrow(simple_hits_amphibia_16s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_16s)

quality_hit_per_asv_16s = nrow(quality_hits_16s) / nrow(simple_asvs_16s)
quality_amphibia_hit_per_asv_16s = nrow(quality_hits_amphibia_16s) / nrow(simple_asvs_16s)

any_quality_hit_per_asv_16s = nrow(quality_hits_16s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_16s)
any_quality_amphibia_hit_per_asv_16s = nrow(quality_hits_amphibia_16s %>%
                             select(asv_id) %>%
                             distinct()) / nrow(simple_asvs_16s)

rarefied_hits = rarefied %>%
  mutate(median_q_hits_vertebrata = recode_values(gene,
                                                "12S" ~ median_asv * quality_hit_per_asv_12s,
                                                "16S" ~ median_asv * quality_hit_per_asv_16s),
         # mean_hits_vertebrata = recode_values(gene,
         #                                        "12S" ~ nn * mean_hit_per_asv_12s,
         #                                        "16S" ~ median_asv * mean_hit_per_asv_16s),
         median_q_hits_amphibia = recode_values(gene,
                                                "12S" ~ median_asv * quality_amphibia_hit_per_asv_12s,
                                                "16S" ~ median_asv * quality_amphibia_hit_per_asv_16s))

ggplot(rarefied_hits, aes(x = nn, y = median_q_hits_vertebrata, color = gene, fill = gene)) +
  geom_line()
  #geom_line(aes(y = mean_seq), linetype = 2)

ggplot(rarefied_hits, aes(x = nn, y = median_q_hits_amphibia, color = gene, fill = gene)) +
  geom_line()
  #geom_line(aes(y = mean_seq), linetype = 2)

```

# summarize relative rates 
```{r}
# ASUs per sample
## at 20 samples:
peace = rarefied %>%
  filter(nn == 20) %>%
  pull(mean_asv_per_sample)
asv_per_samp_12s = peace[1]
asv_per_samp_16s = peace[2]

r_asv_per_samp = asv_per_samp_12s / asv_per_samp_16s

# at least one quality vertebrate hit per ASV
any_quality_hit_per_asv_12s
any_quality_hit_per_asv_16s
r_quality_hit_per_asv = any_quality_hit_per_asv_12s / any_quality_hit_per_asv_16s

# at least one quality amphibia hit per ASV
any_quality_amphibia_hit_per_asv_12s
any_quality_amphibia_hit_per_asv_16s
r_quality_amphibia_hit_per_asv = any_quality_amphibia_hit_per_asv_12s / any_quality_amphibia_hit_per_asv_16s

## classifier pool factor
### prob hit given target amphibian
hope = cumulative_table %>%
  filter(sequences_n == 1) %>%
  pull(p_species_gte_ii)
p_target_amphibian_12s = hope[1]
p_target_amphibian_16s = hope[2]

r_target_amphibian = p_target_amphibian_12s / p_target_amphibian_16s

### prob quality hit given hit (amphibian)
p_quality_hit_12s = nrow(simple_hits_amphibia_12s) / nrow(simple_hits_amphibia_12s %>%
                                        filter(percent_identical >= 95))
p_quality_hit_16s = nrow(simple_hits_amphibia_16s) / nrow(simple_hits_amphibia_16s %>%
                                        filter(percent_identical >= 95))

r_quality_hit = p_quality_hit_12s / p_quality_hit_16s

### prob amphibian given ASV
xx_12s = any_quality_amphibia_hit_per_asv_12s / (p_target_amphibian_12s * p_quality_hit_12s)

xx_16s = any_quality_amphibia_hit_per_asv_16s / (p_target_amphibian_16s * p_quality_hit_16s)

r_xx = xx_12s / xx_16s

```